### PR TITLE
Add decorators

### PR DESCRIPTION
**Pull request by balloob**
Lit, Litter, Litst.

This enables two stage 2 proposals that will make our live easier. I've updated a few classes as examples.

FYI @bramkragten @iantrich @zsarnett @thomasloven 

## Decorators

This will allow us to define class properties and Lit properties in one go, instead of having a `public hass?: HomeAssistant` and a `static get properties() {`. Just add `@property` in front of a property, and Lit will re-render when it's changed.

## class properties

How we're defining types for class properties in TypeScript today, is actually a JS proposal to define and initialize properties. This is required for decorators to work. It will also allow us to initialize instance variables with values inside class definition, as if we are inside the constructor.

```ts
class Bla extends LitElement {
  // combine both things
  @property cache: any = { something: [] };
  
  // just decorate
  @property hass?: HomeAssistant;

  // Just initialize, can be functions too!
  processedInput = memoizeOne((entities: HassEntity[]) => {
    // do some expensive operation
    return result;
  })
}
```

### CODE REVIEW

**Comment by bramkragten** [comment]
Can we also use the `@customElement` decorator now?

**Comment by balloob** [comment]
Yeah, but I got a warning for unused var so ended up not using it. I couldn't get the query one to work 😐

File: src/layouts/app/home-assistant.ts [hunk 1]
```diff
@@ -43,19 +37,9 @@ export class HomeAssistantAppEl extends ext(HassBaseMixin(LitElement), [
   NotificationMixin,
   dialogManagerMixin,
 ]) {
-  private _route?: Route;
-  private _error?: boolean;
-  private _panelUrl?: string;
-
-  static get properties(): PropertyDeclarations {
-    return {
-      hass: {},
```
Comment (bramkragten) [inline src/layouts/app/home-assistant.ts]: Was it intention that `hass` and `routeData` are no longer properties?

Comment (balloob) [reply src/layouts/app/home-assistant.ts]: Yes. Hass is not defined in this file but inside base mixing. Route data was not used anymore

**Pull request merged**